docs/architecture/ADR-003-borrow-assets.md

```diff
@@ -5,106 +5,155 @@
 type MsgSetCollateral struct {
   Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
   Denom    string         `json:"denom" yaml:"denom"`
   Enable   bool           `json:"enable" yaml:"enable"`
 }
+
 // MsgBorrowAsset - a user wishes to borrow assets of an allowed type
 type MsgBorrowAsset struct {
   Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
+  // not a uToken
   Amount   sdk.Coin       `json:"amount" yaml:"amount"`
 }
+
 // MsgRepayAsset - a user wishes to repay assets of a borrowed type
 type MsgRepayAsset struct {
   Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
   Amount   sdk.Coin       `json:"amount" yaml:"amount"`
 }
+
+// MsgSetCollateralMaxUtilization is going to be used as CosmosSDK 0.46 gov message.
+type MsgSetCollateralMaxUtilization {
+  // must be x/gov
+  string signer = 1;
+  string token = 2;
+  // a number between 0 and 10'000 - ratio in basis points.
+  uint32 max_utilization = 3;
+}
 ```
-Messages must use denominations only in the allow-list. Collateral is always a uToken denomination, and assets are never uTokens.
 
-Asset borrowing respects the user's borrowing limit, which compares the total value of a borrower's collateral (scaled by each token's `CollateralWeight`) with the borrower's total borrowed value. Price `x/oracle` must be used to calculate borrow and collateral values.
+Tokens used in above messages must belong to the allow-list. Collateral must be a uToken.
+
+Messages must be signed by the borrower's account.
 
-It is necessary that messages be signed by the borrower's account. Thus the method `GetSigners` should return the `Borrower` address for all message types above.
+Both CLI and gRPC must be supported for the above messages.
 
-### Module State (Keeper) Additions
+### Storage layout
 
-Using the `sdk.Coins` built-in type, which combines multiple {Denom,Amount} pairs as a single object, the `umee/x/leverage` module may remember open borrow and collateral positions as follows:
+Using the `sdk.Coins` built-in type, which combines multiple {Denom,Amount} pairs as a single object, the `umee/x/leverage` module stores open borrow and collateral positions as follows:
 
 ```go
-// pseudocode
-// for each borrower address store open borrows
+// open borrows:
 borrowPrefix | lengthPrefixed(borrowerAddress) | tokenDenom = sdk.Int
-// additionally borrower collateral settings are stored for enabled denoms
-collateralSettingPrefix | lengthPrefixed(borrowerAddress) | tokenDenom = true/false
-// and the amount of collateral stored for each denom is kept
+
+// borrower collateral settings for enabled denoms:
+collateralSettingPrefix | lengthPrefixed(borrowerAddress) | tokenDenom = 0x01
+
+// and the amount of collateral deposited for each uToken:
 collateralAmountPrefix | lengthPrefixed(borrowerAddress) | tokenDenom = sdk.Int
+
+// max token collateral utilization setings
+maxCollateralUtilizationPrefix | token = bigEndian(uint32)
```

Comment:
```suggestion
```
Per-token parameters should be fields in the `Token` struct (stored in registry, updated as a whole by governance)

Reply:
OK